File: javascript/data-structures/p235_Queue.js
/*
    CLRS Section 10.1, p. 235 & Exercise 10.1-4, p. 235
*/

class Queue {
    constructor(Q = [], head = 0, tail = 0) {
        this.Q = Q;
        this.head = head;
        this.tail = tail;
    }

    queueEmpty = () => this.head === this.tail;

    queueFull = () => (this.head === this.tail + 1) 
                        || (this.head === 0 && this.tail === this.Q.length - 1);

    enqueue = (x) => {
        if (this.queueFull()) {
            throw new Error('overflow');
        } else {
            this.Q[this.tail] = x;
            if (this.tail === this.Q.length - 1) {
                this.tail = 0;
            } else {
                this.tail++;
            }
        }
    }

    dequeue = () => {
        if(this.queueEmpty()) {
            throw new Error('underflow');
        } else {
            const x = this.Q.splice(this.head, 1, undefined);
            if (this.head === this.Q.length - 1) {
                this.head = 0;
            } else {
                this.head++;
            }
            return x;
        }
    }
}

module.exports = {
    Queue
}

File: python/data_structures/p235_queue.py
'''
    CLRS Section 10.1, p. 235 & Exercise 10.1-4, p. 235
'''

class Queue:
    def __init__(self, Q=None, head=0, tail=0):
        if Q is None:
            self.Q = []
        else:
            self.Q = Q
        self.head = head
        self.tail = tail
    
    def queue_empty(self):
        return self.head == self.tail
    
    def queue_full(self):
        return (self.head == self.tail + 1) \
            or (self.head == 0 and self.tail == len(self.Q) - 1)

    def enqueue(self, x):
        if self.queue_full():
            raise Exception('overflow')
        else:
            self.Q[self.tail] = x
            if self.tail == len(self.Q) - 1:
                self.tail = 0
            else:
                self.tail += 1
    
    def dequeue(self):
        if self.queue_empty():
            raise Exception('underflow')
        else:
            x = self.Q[self.head]
            self.Q[self.head] = None
            if self.head == len(self.Q) - 1:
                self.head = 0
            else:
                self.head += 1
            return x
